### ISSUE TITLE
`refine` doesn't throw an error until `password` field validation passed

### ISSUE THREAD

**Issue by jorisre** [colinhacks/zod#586]
Hey 👋🏻

I'm playing with Zod 3 and I get different behavior between v1 and v3 about refine.

With V1, `refine` throw an error even if the `password` validation itself isn't satisfied.
👉🏻 https://codesandbox.io/s/epic-dan-fvwtn

Whereas with the v3 the error isn't thrown until the `password` field is good.
👉🏻 https://codesandbox.io/s/gracious-euler-e9jid

Is that new behavior normal?

**Comment by chris-muller** [comment]
For additional information in case anyone is looking into this, I had a `refine` on an object shape and the refine wouldn't get run until all previous parsing succeeded, even properties not being used inside the `refine` method.

This caused issues for me because some of the validation errors would only appear after most of the form was completed.

**Comment by scotttrinh** [comment]
Yeah, this is expected behavior in v3. The refinement (and other "effects" like transformation) happen after the parsing, and the parsing will halt the whole pipeline if it encounters errors in the parsing step.

For more details on the motivation behind this part of the refactor, see #264
